### PR TITLE
Pyrrha Localization : Language bo_CN

### PR DESCRIPTION
**Pull request by Lungsang-bod**

### CODE REVIEW

File: app/templates/main/tokens_del_row.html [hunk 1]
```diff
@@ -4,7 +4,7 @@
 {% block content %}
 
 <form method="POST" action="{{ url_for('.tokens_del_row', token_id=token.id, corpus_id=corpus.id)}}" class="form col-md-9">
-    <h1>Corpus {{corpus.name}} - Remove a token</h1>
+    <h1>{{ _('Corpus') }} {{corpus.name}} - Remove a token</h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_del_row.html:7]: Shouldn't `Remove a token` be translated too ?

File: app/templates/main/tokens_del_row.html [hunk 2]
```diff
@@ -4,7 +4,7 @@
...
     <p>This page allows you to remove the following one</p>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_del_row.html:8]: Missing translation

File: app/templates/main/tokens_edit_form.html [hunk 3]
```diff
@@ -4,7 +4,7 @@
 {% block content %}
 
 <form method="POST" action="{{ url_for('.tokens_edit_form', token_id=token.id, corpus_id=corpus.id)}}" class="form col-md-9">
-    <h1>Corpus {{corpus.name}} - Token Edit</h1>
+    <h1>{{ _('Corpus') }} {{corpus.name}} - Token Edit</h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_edit_form.html:7]: Missing translation for `Token Edit`

File: app/templates/main/tokens_edit_form.html [hunk 4]
```diff
@@ -4,7 +4,7 @@
...
     <p>This page allows you to correct the form of the following token</p>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_edit_form.html:8]: Missing translation for whole paragraph

File: app/templates/main/tokens_edit_history.html [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 {% block content %}
 
-<h1>Corpus {{corpus.name}} - History of tokens edition</h1>
+<h1>{{ _('Corpus') }} {{corpus.name}} - History of tokens edition</h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_edit_history.html:6]: Missing translation for `History of tokens edition`

File: app/templates/main/tokens_export.html [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 {% block content %}
 
-<h1>Corpus {{corpus.name}} - Download </h1>
+<h1>{{ _('Corpus') }} {{corpus.name}} - Download </h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_export.html:5]: Missing translation for `Download`

File: app/templates/main/tokens_history.html [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 {% block content %}
 
-<h1>Corpus {{corpus.name}} - List of tokens </h1>
+<h1>{{ _('Corpus') }} {{corpus.name}} - List of tokens </h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_history.html:7]: Missing translation for `List of tokens`

File: app/templates/main/tokens_search_through_fields.html [hunk 8]
```diff
@@ -7,7 +7,7 @@
 
 {% set visible = corpus.get_columns_headings() %}
 
-    <h1 class="mt-3"><i class="fa fa-search mr-2"></i>Corpus {{ corpus.name }} - Search tokens within the corpus </h1>
+    <h1 class="mt-3"><i class="fa fa-search mr-2"></i>{{ _('Corpus') }} {{ corpus.name }} - Search tokens within the corpus </h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_search_through_fields.html:10]: Missing translation for `Search tokens within the corpus`

File: app/templates/main/tokens_similar_to_record.html [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 {% set visible = corpus.displayed_columns_by_name %}
 
-<h1>Corpus {{corpus.name}} - List of tokens </h1>
+<h1>{{ _('Corpus') }} {{corpus.name}} - List of tokens </h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_similar_to_record.html:10]: Missing translation for `List of tokens`

File: app/templates/main/tokens_similar_to_token.html [hunk 10]
```diff
@@ -7,7 +7,7 @@
 
 {% set visible = corpus.displayed_columns_by_name %}
 
-<h1>Corpus {{corpus.name}} - Similar tokens </h1>
+<h1>{{ _('Corpus') }} {{corpus.name}} - Similar tokens </h1>
```
Comment (PonteIneptique) [inline app/templates/main/tokens_similar_to_token.html:10]: Missing translation for `Similar tokens`

File: app/templates/macros/nav_macros.html [hunk 11]
```diff
@@ -204,50 +204,50 @@
...
         {% endif %}
     </div>
     <div class="menu-horizontal-group mx-1">
         <a class="btn btn-outline-dark btn-sm" title="Quickly return to a bookmarked token" id="bookmark_link" href="{{ url_for("main.corpus_bookmark", corpus_id=corpus.id) }}">
-            <i class="fa fa-bookmark"></i> Go to bookmark
+            <i class="fa fa-bookmark"></i> {{ _('Go to bookmark') }}
         </a>
         <div class="btn-group btn-group-sm" role="group">
             <a class="btn btn-outline-dark title">
                 Annotations
```
Comment (PonteIneptique) [inline app/templates/macros/nav_macros.html:221]: Missing translation for `Annotations`

**Review by PonteIneptique — CHANGES_REQUESTED** [review]
I spotted many non translated things. Is it intentional ?

**Comment by Lungsang-bod** [comment]
> I spotted many non translated things. Is it intentional ?

Yes @PonteIneptique , there are still lots of non translated things because right now I dont have the exact translation word for them.   But if it's okay, i can mark and extract all the remaining translatable strings in message.po

**Comment by PonteIneptique** [comment]
> > I spotted many non translated things. Is it intentional ?
> 
> Yes @PonteIneptique , there are still lots of non translated things because right now I dont have the exact translation word for them. But if it's okay, i can mark and extract all the remaining translatable strings in message.po :)

Yup, I think it would be good to at least have everything in extracted, even if the translation is incomplete. This will allow for only working on `message.po` in the future :)

**Comment by Lungsang-bod** [comment]
> > > I spotted many non translated things. Is it intentional ?
> > 
> > 
> > Yes @PonteIneptique , there are still lots of non translated things because right now I dont have the exact translation word for them. But if it's okay, i can mark and extract all the remaining translatable strings in message.po :)
> 
> Yup, I think it would be good to at least have everything in extracted, even if the translation is incomplete. This will allow for only working on `message.po` in the future :)

Okay ! i am on it :)

File: app/templates/main/corpus_info.html [hunk 12]
```diff
@@ -54,7 +54,7 @@ <h2>{{ _('Statistics') }}</h2>
                             <text x="18" y="20.35" class="percentage">{{'%0.2f' % (100 - stats.lemma_acc)}}%</text>
                         </svg>
                         <p class="stats-info">
-                        {{ _('lemma were correct from lemmatization. It means you corrected {{stats.lemma_count}} lemma.') }}
+                        {{ _('lemma were correct from lemmatization. It means you corrected') }} {{stats.lemma_count}} {{ _('lemma.') }}
```
[the remark this answers is not in the document]
Comment (Lungsang-bod) [reply app/templates/main/corpus_info.html:57]: Okay :) i will try this one. thank you !

[the remark this answers is not in the document]
Comment (Lungsang-bod) [reply app/templates/main/corpus_info.html:57]: I think above two methods are used for extracting from python source code. Not from template. I tried them in template and it's not extracting :)

[the remark this answers is not in the document]
Comment (PonteIneptique) [reply app/templates/main/corpus_info.html:57]: Hmm, I'll have a look at the end of the PR for this kinda things, I also want to add a user pref :)

[the remark this answers is not in the document]
Comment (Lungsang-bod) [reply app/templates/main/corpus_info.html:57]: ```<p>{{ _('Alternatively, you can paste the following link in your browser's address bar:') }}</p>```
When extracting this sentence, the browser's (') is causing a problem.

```<p>The {{ config.APP_NAME }} Team</p>```

```<p>To reset your password, <a href="{{ reset_link }}">click here</a>.</p>```

There are many sentences like above three and I am figuring out how to extract the whole sentence without to break them up. Please let me know if you have any solution :)

[the remark this answers is not in the document]
Comment (Lungsang-bod) [reply app/templates/main/corpus_info.html:57]: ```annotations <br />corrected```
Including these kind, where </br> is put between a sentence.

[the remark this answers is not in the document]
Comment (Lungsang-bod) [reply app/templates/main/corpus_info.html:57]: > Hmm, I'll have a look at the end of the PR for this kinda things, I also want to add a user pref :)

Okay :)

**Comment by PonteIneptique** [comment]
Hey @Lungsang-bod , I was actually waiting for your "go" :)
I don't know if I'll have the time to review and correct by next week, but I'll try :)

**Comment by Lungsang-bod** [comment]
Thank you @PonteIneptique :)

**Comment by Lungsang-bod** [comment]
Sure @PonteIneptique, in the meantime we will push the translated strings in the PR. Please have a look when time permits you. 
Wishing the best for your new work :)

**Comment by PonteIneptique** [comment]
I'll try to deal with this tomorrow, if I can. By the way, have you and your team published any paper I could see about what you do with Pyrrha ? :)

**Comment by PonteIneptique** [comment]
@Lungsang-bod I reworked your PR in https://github.com/hipster-philology/pyrrha/pull/285 so that it includes the necessary bits.

Can you check that it works for you ?

By the way, if you need a variable in a template, you can use the following snippet:

```html
{{_("lemma were correct from lemmatization. It means you corrected %(count)s lemma.", count=stats.lemma_count)}}
```

https://github.com/hipster-philology/pyrrha/blob/0f6b01f4ee171affb3f20f01691bb54fead943d1/app/templates/main/corpus_info.html#L57

For future message upgrades, please check the readme.md but basically run 

```sh
python manage.py translate update
# Change the translation and then do
python manage.py translate compile
```

and please commit with prefix `(i18n/bo_CN) and then your message`

Finally, you might need to run

```sh
python manage.py db-upgrade user-language
```

as I modified the models :)

**Comment by Lungsang-bod** [comment]
> I'll try to deal with this tomorrow, if I can. By the way, have you and your team published any paper I could see about what you do with Pyrrha ? :)

I will ask my team if any paper is published regarding Pyrrha and will let you know :)

**Pull request merged**